service/dap/server.go

```diff
@@ -640,38 +657,37 @@ func isValidLaunchMode(launchMode interface{}) bool {
 // (in our case this TCP server) can be terminated.
 func (s *Server) onDisconnectRequest(request *dap.DisconnectRequest) {
 	s.send(&dap.DisconnectResponse{Response: *newResponse(request.Request)})
+	s.mu.Lock()
 	if s.debugger != nil {
-		_, err := s.debugger.Command(&api.DebuggerCommand{Name: api.Halt}, nil)
-		if err != nil {
-			switch err.(type) {
-			case *proc.ErrProcessExited:
-				s.log.Debug(err)
-			default:
-				s.log.Error(err)
-			}
-		}
-		// We always kill launched programs
-		kill := s.config.Debugger.AttachPid == 0
+		// We always kill launched programs.
 		// In case of attach, we leave the program
-		// running but default, which can be
+		// running by default, which can be
 		// overridden by an explicit request to terminate.
-		if request.Arguments.TerminateDebuggee {
-			kill = true
-		}
-		err = s.debugger.Detach(kill)
-		if err != nil {
-			switch err.(type) {
-			case *proc.ErrProcessExited:
-				s.log.Debug(err)
-			default:
-				s.log.Error(err)
-			}
-		}
+		killProcess := s.config.Debugger.AttachPid == 0 || request.Arguments.TerminateDebuggee
+		s.stopDebugSession(killProcess)
 	} else {
 		s.stopNoDebugProcess()
 	}
-	// TODO(polina): make thread-safe when handlers become asynchronous.
-	s.signalDisconnect()
+	s.mu.Unlock()
```

Comment:
Will the following work? (Sorry - it's hard for me to use github codereview ui :-( )

defer s.triggerServerStop()
s.mu.Lock()
defer s.mu.Unlock()
if s.debugger != nil {
   ... 
   s.stopDebugSession(killProcess)
 } else {
   s.stopNodebugProcess()
 }

Reply:
done